test/README.md

```diff
@@ -1,95 +1,284 @@
+
+# The Generic MQTT tester image follows a slightly different build pattern. Use the following
+# steps to build and push this image to the local container repository.
+sudo scripts/linux/cross-platform-rust-build.sh --os ubuntu18.04 --arch amd64 --build-path test/modules/generic-mqtt-tester/
+cd test/modules/generic-mqtt-tester
+sudo docker build --no-cache -t localhost:5000/microsoft/generic-mqtt-tester:latest-linux-amd64 --file docker/linux/amd64/Dockerfile --build-arg EXE_DIR=. target
+sudo docker push localhost:5000/microsoft/generic-mqtt-tester:latest-linux-amd64
+~~~
+
+### Test
+
+#### Test parameters
+
+The end-to-end tests take several parameters, which they expect to find in a file named `context.json` in the same directory as the test binaries (e.g., `test/Microsoft.Azure.Devices.Edge.Test/bin/Debug/netcoreapp2.1/context.json`). Parameter names are case-insensitive. See [end-to-end test parameters](./doc/end-to-end-test-config) for details. 
+
+#### Sample context.json
+~~~ json
+{
+"packagePath": "/home/azureuser/iotedge/edgelet/target/release/",
+"edgeAgentImage": "localhost:5000/microsoft/edge-agent:latest-linux-amd64",
+"edgeHubImage": "localhost:5000/microsoft/edge-hub:latest-linux-amd64",
+"loadGenImage": "localhost:5000/microsoft/load-gen:latest-linux-amd64",
+"methodSenderImage": "localhost:5000/microsoft/direct-method-sender:latest-linux-amd64",
+"methodReceiverImage": "localhost:5000/microsoft/direct-method-receiver:latest-linux-amd64",
+"networkControllerImage": "localhost:5000/microsoft/network-controller:latest-linux-amd64",
+"relayerImage": "localhost:5000/microsoft/relayer:latest-linux-amd64",
+"tempFilterFuncImage": "localhost:5000/microsoft/temperature-filter-function:latest-linux-amd64",
+"tempFilterImage": "localhost:5000/microsoft/temperature-filter:latest-linux-amd64",
+"tempSensorImage": "localhost:5000/microsoft/simulated-temperature-sensor:latest-linux-amd64",
+"testResultCoordinatorImage": "localhost:5000/microsoft/test-result-coordinator:latest-linux-amd64",
+"numberLoggerImage": "localhost:5000/microsoft/number-logger:latest-linux-amd64",
+"genericMqttTesterImage": "localhost:5000/microsoft/generic-mqtt-tester:latest-linux-amd64",
+"diagnosticsImage": "localhost:5000/microsoft/diagnostics:latest-linux-amd64",
+"metricsValidatorImage": "localhost:5000/microsoft/metrics-validator:latest-linux-amd64",
+"caCertScriptPath": "/home/azureuser/iotedge/test/certs",
+"dpsIdScope": "0ne01F35169",
+"rootCaCertificatePath": "/home/azureuser/iotedge/test/certs/certs/azure-iot-test-only.root.ca.cert.pem",
+"rootCaPrivateKeyPath": "/home/azureuser/iotedge/test/certs/private/azure-iot-test-only.root.ca.key.pem",
+"verbose": "true",
+"logFile": "/home/azureuser/iotedge/logs/logfile",
+"registries":
+        [
+                {
+                        "address": "localhost:5000",
+                        "username": "username"
+                }
+        ]

```

Comment:
Can you indent the body of this JSON object? Also, can you do 4-space indent instead of 8-space?

Reply:
Will do